### ISSUE TITLE
Prompt evaluation performance regression in llama.cpp on RDNA3 with HSA_OVERRIDE_GFX_VERSION=11.0.1 vs 11.0.0

### ISSUE THREAD

**Issue by Googulator** [ROCm/ROCm#2590]
We're evaluating llama.cpp with ROCm offloading on various RDNA3 GPUs (primarily RX 7800 XT and RX 7900 XT).

On initial testing, we found that a 13b model with Q6_K quantization fully offloaded to GPU (-ngl 43) showed significantly slower prompt evaluation times on the 7800 compared to the 7900, far beyond what would be expected from the relative performance difference between these 2 GPUs. On the 7800, prompt evaluation would take more than a minute on our example prompt, while it was near instantaneous (less than a second) on the 7900.

Upon further investigation, the RX 7800 shows a 3 second penalty for every 64 tokens of prompt (0-64 tokens 3 seconds, 65-128 tokens 6s, 129-192 tokens 9s, and so on) over the 7900.

Since the 7800 is recognized by ROCm as gfx1101, vs gfx1100 on the 7900, we tried setting HSA_OVERRIDE_GFX_VERSION=11.0.0, which led to massive performance improvement, with the 7800 only proportionally slower than the 7900, as expected based on the specifications. We then tested HSA_OVERRIDE_GFX_VERSION=11.0.1 (and 11.0.2) on the 7900, and saw the exact same performance issue as on the 7800.

The testing was performed on Ubuntu 22.04.3 Server, using ROCm v5.7.0, apparently the latest version available for this version of Ubuntu.

**Comment by Pierre-vh** [comment]
Hi,

We're investigating this, can you provide some instructions to reproduce this issue? 
How can we test llama.cpp in the same way you did?

Thanks

**Comment by kzhuravl** [comment]
@Googulator, gentle ping. Can you provide repro steps?

**Comment by Googulator** [comment]
Hi!

You will need a 7800 XT, 7900 GRE, 7900 XT or 7900 XTX card, the latest llama.cpp code, and a matching recent Llama 2-13B-based GGUF model.

Export HSA_OVERRIDE_GFX_VERSION=11.0.1 in your environment, then compile llama.cpp for ROCm (using the LLAMA_HIPBLAS=1 option).

Launch the server:
```
./server -m /path/to/model.gguf -c 4096 -ngl 43 --mlock --host 0.0.0.0 --port 8080
```

Open http://localhost:8080 in your browser, set temperature to zero, and insert the following prompt in the bottom field:

```
What happened to Apollo 13?
```

(Don't change any other options.)

Click Send, then measure the time it takes before Llama starts generating text. (The speed of text generation is unaffected, it's the time it takes for Llama to even begin generating text that takes time. Also, only the first invocation with each prompt is slow - subsequent invocations without restarting the server will come from the prompt cache, and thus hide the regression.)

Now, repeat this procedure with HSA_OVERRIDE_GFX_VERSION=11.0.0 exported in your environment (instead of 11.0.1). Observe how the text starts to get generated almost instantaneously,instead of a multi-second delay as with 11.0.1.

If you replace the prompt with something longer (e.g. "What is this song called?" followed by the lyrics of your country's national anthem), the difference gets exponentially worse.

**Comment by SlyEcho** [comment]
Hello, I was one of the people that worked on ROCm support on llama.cpp. While the code that we came up with in the end is working, it is still not really well tested on all GPUs.

I would say that there is still an issue with ROCm itself because of lacking certain features that CUDA has namely the `__dp4a` integer intrinsic. I think it would be much better if it were implemented by 1st party engineers from AMD.
